src/uint256.h

```diff
@@ -89,6 +89,14 @@ class base_blob
                ((uint64_t)ptr[7]) << 56;
     }
 
+    std::array<uint8_t, WIDTH> GetRawBytes() const
+    {
+        std::array<uint8_t, WIDTH> buf = {};
+        memcpy(buf.begin(), (void*)data, WIDTH);
```

Comment:
Huh, this shouldn't work. [`std::array::begin` returns a C++ iterator](https://en.cppreference.com/w/cpp/container/array/begin), which is not the same thing as a pointer to the underlying data (which you get from `std::array::data`), but [`std::memcpy` reinterprets its arguments as arrays of `unsigned char`](https://en.cppreference.com/w/cpp/string/byte/memcpy). I'd have expected this to fail with a compile error, and yet it compiles and passes tests. Maybe a clang thing?

I'd have instead written this as:
```suggestion
        memcpy(buf.data(), this->begin(), WIDTH);
```
(because `uint256::begin` is unfortunately actually implementing what `::data` should do), or use `std::copy` instead which is aware of C++ iterators (but likely slower):
```suggestion
        memcpy(this->begin(), this->end(), buf.begin());
```
(which works with both iterators and plain pointers).

Reply:
Aha, it works because `iterator` is indeed just `value_type*` in libc++:
```cpp
template <class _Tp, size_t _Size>
struct _LIBCPP_TEMPLATE_VIS array
{
    // types:
    typedef array __self;
    typedef _Tp                                   value_type;
    typedef value_type&                           reference;
    typedef const value_type&                     const_reference;
    typedef value_type*                           iterator;
    typedef const value_type*                     const_iterator;
```
and the iterators get this type directly from `::data`:
```cpp
    // iterators:
    _LIBCPP_INLINE_VISIBILITY _LIBCPP_CONSTEXPR_AFTER_CXX14
    iterator begin() _NOEXCEPT {return iterator(data());}
    _LIBCPP_INLINE_VISIBILITY _LIBCPP_CONSTEXPR_AFTER_CXX14
    const_iterator begin() const _NOEXCEPT {return const_iterator(data());}
```
and here `_Tp` is `uint8_t`, so after all that `buf.begin()` returns `uint8_t*` which indeed can be reinterpreted as `unsigned char*`. It's a bit convoluted, so switching to `buf.data()` would remove the indirection, but there's no bug here.

Reply:
Well that's dependent on an implementation detail of libc++, right? I would call that a bug.

(Note that you didn't change the call to `std::copy` in your second suggestion. I prefer the first, anyway.)

Reply:
Agreed, I can't find a standard that guarantees this will be correct, so we should fix this. Reference: https://en.cppreference.com/w/cpp/container/array and https://en.cppreference.com/w/cpp/named_req/RandomAccessIterator